src/lib/reasoners/adt_rel.ml

```diff
@@ -798,41 +619,72 @@ let assume env uf la =
       Printer.print_dbg ~module_name:"Adt_rel" ~function_name:"assume"
         "assume deduced %d equalities@ %a" (List.length eqs)
         (Printer.pp_list_no_space print) eqs;
-    env, { Sig_rel.assume = eqs; remove = [] }
+    let eqs = List.rev_append eqs result.assume in
+    { env with delayed }, { Sig_rel.assume = eqs; remove = [] }
 
 
 (* ################################################################ *)
 
 let two = Numbers.Q.from_int 2
 
+let constr_of_destr ty d =
+  match ty with
+  | Ty.Tadt (name, params) ->
+    begin
+      let Ty.Adt cases = Ty.type_body name params in
+      try
+        let r =
+          List.find
+            (fun Ty.{ destrs; _ } ->
+               List.exists (fun (d', _) -> Hstring.equal d d') destrs
+            ) cases
+        in
+        r.constr
+      with Not_found -> assert false
+    end
+
+  | _ -> assert false
+
+exception Found of X.r * Hstring.t
+
+(* Pick a delayed destructor application in [env.delayed].
+
+   @raise Found (r, d) if there is a delayed application of the destructor
+          [d] on the semantic value [r]. *)
+let pick_delayed_destructor env =
```

Comment:
Not a big fan of returning values through an exception, I personally wouldn't mind paying the cost of a `try .. with` to return an option here.

Reply:
Actually I hesitated between the two versions. I will rollback to the previous one.